[flang] flang manpage overhaul (#144948)
llvm/llvm-project: flang/docs/CommandGuide/index.rst

flang - the Fortran compiler
============================

SYNOPSIS
--------

:program:`flang` [*options*] *filename ...*

DESCRIPTION
-----------

> :program:`flang` is a Fortran compiler which encompasses preprocessing, parsing,

Reviewer: This information about the different stages of compilation are interesting, but I imagine most users would already be familiar with compiler drivers with these features (e.g. classic flang can do all of this). I think maybe the description should start with a very short description of what the compiler supports: something like
> Flang supports all of Fortran 95 and many newer language features. Flang supports OpenMP and has some support for OpenACC and CUDA.

I have been deliberately vague here because it could be easy to forget to update this documentation when new features are added. @clementval please could you advise on rough status of OpenACC and CUDA in upstream flang?

It might also be worth giving some very rough getting started information: e.g. observing that with some exceptions we support very similar flags to gfortran and classic-flang.

I expect these two points would give the information that most Fortran programmers need to get started without them having to read much further. Maybe an example of how to compile a very simple fortran program would be beneficial too
> flang [-fopenmp] file1.f90 file2.f90 -O3 -o program

> optimization, code generation, assembly, and linking.  Depending on which
> high-level mode setting is passed, Flang will stop before doing a full link.

Reviewer 2: ```suggestion
optimization, code generation, assembly, and linking.  Depending on the
options passed in, Flang will perform only some, or all, of these actions.
```

Author: Done

While Flang is highly integrated, it is important to understand the stages of
compilation, to understand how to invoke it.  These stages are:

Driver

>     The flang executable is actually a small driver which controls the overall

Author: Done

    execution of other tools such as the compiler, assembler and linker.
    Typically you do not need to interact with the driver, but you
    transparently use it to run the other tools.

Preprocessing
    This stage handles tokenization of the input source file, macro expansion,

>     #include expansion and handling of other preprocessor directives.

Author: Done

Parsing and Semantic Analysis
    This stage parses the input file, translating preprocessor tokens into a
    parse tree.  Once in the form of a parse tree, it applies semantic

>     analysis to compute types for expressions as well and determine whether

Author: Done

    the code is well formed. This stage is responsible for generating most of

>     the compiler warnings as well as parse errors.

Author: Done

Code Generation and Optimization
    This stage creates the intermediate code (known as "LLVM IR") and ultimately
    to machine code.  This phase is responsible for optimizing the generated
    code and handling target-specific code generation. The output of this stage
    is typically called a ".s" file or "assembly" file.

    Flang also supports the use of an integrated assembler, in which the code
    generator produces object files directly. This avoids the overhead of

>     generating the ".s" file and of calling the target assembler.

Author: Done

Assembler
    This stage runs the target assembler to translate the output of the
    compiler into a target object file. The output of this stage is typically
    called a ".o" file or "object" file.

Linker
    This stage runs the target linker to merge multiple object files into an

>     executable or dynamic library. The output of this stage is typically called

Author: Done
